src/injectors-cache.ts

```diff
@@ -0,0 +1,148 @@
+    : cache.match(cacheKey(url, key));
+}
+
+/**
+ * Saves an injected Response to the cache twice, for raw and Brotli-compressed.
+ *
+ * @param response - raw Response to save to cache.
+ * @param url - base file URL that has dynamic content injected into it.
+ * @param key - unique identifiers for this specific injection.
+ */
+async function saveCache(
+  response: Response,
+  url: string,
+  key: string
+): Promise<void> {
+  console.log('Caching', url, 'with cache key', key);
+  if (!response.body) {
+    throw new Error('Response has no body');
+  }
+
+  // Don't cache short-lived cache controls.
+  response.headers.delete(HeaderKeys.CACHE_CONTROL);
+
+  // Tee the body as it is going to be read twice - once for saving the plain
+  // response and once for performing Brotli compression.
+  const [body1, body2] = response.body.tee();
+  const plainResponse = new Response(body1, response);
+
+  const cache = await openInjectorsCache();
+  const plainResponsePromise = cache
+    .put(cacheKey(url, key), plainResponse)
+    .then(() => {
+      console.log('> Plain response cached');
+    });
+
+  // Create an equivalent Response object that is already Brotli-compressed.
+  const bodyAsText = await new Response(body2, response).text();
+  const compressed = await brotli.compress(textEncoder.encode(bodyAsText), {
+    quality: 11,
+  });
+  const brotliResponse = new Response(compressed, {
+    headers: [
+      ...response.headers,
+      [HeaderKeys.CONTENT_ENCODING, ContentEncoding.BROTLI],
+    ],
+    encodeBody: 'manual',
+  });
+  const brotliResponsePromise = cache
+    .put(cacheKeyBrotli(url, key), brotliResponse)
+    .then(() => {
+      console.log('> Brotli response cached');
+    });
+
+  // Cache both the plain and the Brotli-compressed responses.
+  await Promise.all([plainResponsePromise, brotliResponsePromise]);
+}
+
+/**
+ * Enqueues a cache-save action for after the request, and respond with a clone
+ *
+ * @param extend - wrapper for the FetchEvent's waitUntil method.
+ * @param response - raw Response to save to cache.
+ * @param url - base file URL that has dynamic content injected into it.
+ * @param key - unique identifiers for this specific injection.
+ * @returns a clone of `response`.
+ */
+export async function enqueueCacheAndClone(
+  extend: FetchEvent['waitUntil'],
+  response: Response,
+  url: string,
+  key: string
+): Promise<Response> {
+  if (!response.body) {
+    // To pass type checking below...
+    throw new Error('Response has no body');
+  }
+
+  // If we need to cache this response, we must call .tee on the body as it is
+  // going to be read multiple times.
+  const [body1, body2] = response.body.tee();
+
+  // Do *not* await on this promise. The `waitUntil` call extends the
+  // execution of this call until this promise is resolved, but we want to
+  // respond before the caching is performed.
+  extend(saveCache(new Response(body2, response), url, key));
```

Comment:
```suggestion
  // If we need to cache this response, we must call .clone as it is
  // going to be read multiple times.
  // Do *not* await on this promise. The `waitUntil` call extends the
  // execution of this call until this promise is resolved, but we want to
  // respond before the caching is performed.
  extend(saveCache(response.clone(), url, key));
```

Reply:
That was the first thing I tried, but it turns out that this exhausts `response.body`, which can only be read once (again, `ReadableStream`...). Basically, `response.<anything method that will read the body>` makes the same `response` useless for further reads (its headers can still be used to initialize a new Response, as the 2nd parameter for `new Response`)

Reply:
Expanded on the comment to explains this bit to a fault 492d57d

Reply:
[Oh, turns out .clone does .tee behind the scenes!](https://github.com/ampproject/cdn-worker/pull/96/commits/04e505cddc956b0481fc493b213fe20091a6218b)